target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts

```diff
@@ -0,0 +1,338 @@
+			reg = <2>;
+			phy-mode = "gmii";
+			phy-handle = <&phy11>;
+		};
+
+		ethernet@4 {
+			compatible = "lantiq,xrx200-pdi-port";
+			reg = <4>;
+			phy-mode = "gmii";
+			phy-handle = <&phy13>;
+		};
+
+		ethernet@5 {
+			compatible = "lantiq,xrx200-pdi-port";
+			reg = <5>;
+			phy-mode = "rgmii";
+
+			fixed-link {
+				speed = <1000>;
+				full-duplex;
+			};
+		};
+	};
+
+	mdio {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		compatible = "lantiq,xrx200-mdio";
+
+		phy0: ethernet-phy@0 {
+			reg = <0x0>;
+			compatible = "ethernet-phy-ieee802.3-c22";
+		};
+
+		phy1: ethernet-phy@1 {
+			reg = <0x1>;
+			compatible = "ethernet-phy-ieee802.3-c22";
+		};
+
+		phy11: ethernet-phy@11 {
+			reg = <0x11>;
+			compatible = "lantiq,phy11g", "ethernet-phy-ieee802.3-c22";
+		};
+
+		phy13: ethernet-phy@13 {
+			reg = <0x13>;
+			compatible = "lantiq,phy11g", "ethernet-phy-ieee802.3-c22";
+		};
+	};
+};
+
+&gphy0 {
+	lantiq,gphy-mode = <GPHY_MODE_GE>;
+};
+
+&gphy1 {
+	lantiq,gphy-mode = <GPHY_MODE_GE>;
+};
+
+&gpio {
+	pinctrl-names = "default";
+	pinctrl-0 = <&state_default>;
+
+	state_default: pinmux {
+		phy-rst {
+			lantiq,pins = "io32", "io44";
+			lantiq,pull = <0>;
+			lantiq,open-drain;
+			lantiq,output = <1>;
+		};
+
+		pcie-rst {
+			lantiq,pins = "io21";
+			lantiq,open-drain;
+			lantiq,output = <1>;
+		};
+	};
+
+	pins_spi_default: pins_spi_default {
+		spi_in {
+			lantiq,groups = "spi_di";
+			lantiq,function = "spi";
+		};
+
+		spi_out {
+			lantiq,groups = "spi_do", "spi_clk",
+				"spi_cs4";
+			lantiq,function = "spi";
+			lantiq,output = <1>;
+		};
+	};
+};
+
+&spi {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pins_spi_default>;
+
+	flash@4 {
+		compatible = "jedec,spi-nor";
+		reg = <4>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			urlader: partition@0 {
+				reg = <0x0 0x20000>;
+				label = "urlader";
+				read-only;
+			};
+
+			partition@20000 {
+				reg = <0x20000 0x10000>;
+				label = "tffs (1)";
+				read-only;
+			};
+
+			partition@30000 {
+				reg = <0x30000 0x10000>;
+				label = "tffs (2)";
+				read-only;
+			};
+		};
+	};
+};
+
+&localbus {
+	flash@1 {
+		compatible = "lantiq,nand-xway";
+		bank-width = <2>;
+		reg = <1 0x0 0x2000000>;
+
+		nand-ecc-mode = "on-die";
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "kernel";
+				reg = <0x0 0x400000>;
+			};
+
+			partition@400000 {
+				label = "ubi";
+				reg = <0x400000 0x7c00000>;
+			};
+		};
+	};
+};
+
+&usb_phy0 {
+	status = "okay";
+};
+
+&usb_phy1 {
+	status = "okay";
+};
+
+&usb0 {
+	status = "okay";
+	vbus-supply = <&usb0_vbus>;
+};
+
+&usb1 {
+	status = "okay";
+	vbus-supply = <&usb1_vbus>;
+};
+
+&pcie0 {
+	status = "okay";
+	gpio-reset = <&gpio 21 GPIO_ACTIVE_LOW>;
+	gpio-reset-dev = <&gpio 22 GPIO_ACTIVE_LOW>;
```

Comment:
do you know what this gio-reset-dev pin is used for?
as far as I understand there is only one reset line in the PCIe connector spec, called WAKE# or PERST#: https://en.wikipedia.org/wiki/PCI_Express#Pinout

Reply:
I don't know what it is for, but if it is dropped, the PCIe bus does not respond at all. The same is true for the 3490 (there is an xhci controller on PCIe instead of WiFi). I found it in AVM kernel sources, they toggled it together with the reset pins (as the corresponding kernel patch now does).

Reply:
> In theory this means that we would have to add support for managing the reset-gpios inside the wifi@ node to the ath9k driver.
> But I don't know if that will work in reality, because I think (based on your description) the ath9k driver is only loaded after that reset line is deasserted (= AR9580 is taken out of reset).
I forgot to mention:
the USB subsystem has a similar problem and there's a workaround (called gpio-hog) which is accepted upstream, see here for an example: https://github.com/torvalds/linux/blob/6b14dd7267126931e9a95c68a442e8f2dabdc3c4/arch/arm/boot/dts/meson8b-odroidc1.dts#L285

Reply:
Good points, thanks for the research!
I have a similar problem on the 3490 as well and it seems that the ath79 target on the 3490 needs a reset line for the ath10k chip as well. 
Instead of the gpio-hog it might be sufficient to have a gpio-export for the reset line to force it high, I'll investigate that further and report back. It will take a few days, since I'm a bit busy with work.

Reply:
It works using gpio-hog, I removed the patch and added the gpio-hog to the .dts. Took me quite some time to figure out that gpio-ranges is required, otherwise it fails with -517 (see https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.14/080-pinctrl-msm-fix-gpio-hog-related-boot-issues.patch for the same problem on a different target). I fixed it for all vr9 boards.

Fun fact: my first intention was to simply set the GPIO to a predefined state, but since I wasn't aware of gpio-hog, I came up with the patch.

Adding support for the reset to ath9k wouldn't fix the issue: the PCIe link times out during initialization if the GPIO is not correctly set up, setting it in ath9k would be too late.
```
[    0.430881] ifx_pcie_wait_phy_link_up timeout
[    0.651469] ifx_pcie_wait_phy_link_up timeout
[    0.868731] ifx_pcie_wait_phy_link_up timeout
[    1.086316] ifx_pcie_wait_phy_link_up timeout
[    1.303914] ifx_pcie_wait_phy_link_up timeout
[    1.308149] pcie_rc_initialize link up failed!!!!!
```